cpp-ch/local-engine/Functions/SparkFunctionCheckDecimalOverflow.cpp

```diff
@@ -192,20 +190,62 @@ class FunctionCheckDecimalOverflow : public IFunction
     static bool convertToDecimalImpl(
-        const FromFieldType & decimal, UInt32 precision_to, UInt32 scale_from, UInt32 scale_to, typename ToDataType::FieldType & result)
+        const FromDataType::FieldType & value, UInt32 precision_to, UInt32 scale_from, UInt32 scale_to, typename ToDataType::FieldType & result)
     {
+        using FromFieldType = typename FromDataType::FieldType;
         if constexpr (std::is_same_v<FromFieldType, Decimal32>)
-            return convertDecimalsImpl<DataTypeDecimal<Decimal32>, ToDataType>(decimal, precision_to, scale_from, scale_to, result);
-
+            return convertDecimalsImpl<DataTypeDecimal<Decimal32>, ToDataType>(value, precision_to, scale_from, scale_to, result);
         else if constexpr (std::is_same_v<FromFieldType, Decimal64>)
-            return convertDecimalsImpl<DataTypeDecimal<Decimal64>, ToDataType>(decimal, precision_to, scale_from, scale_to, result);
+            return convertDecimalsImpl<DataTypeDecimal<Decimal64>, ToDataType>(value, precision_to, scale_from, scale_to, result);
         else if constexpr (std::is_same_v<FromFieldType, Decimal128>)
-            return convertDecimalsImpl<DataTypeDecimal<Decimal128>, ToDataType>(decimal, precision_to, scale_from, scale_to, result);
+            return convertDecimalsImpl<DataTypeDecimal<Decimal128>, ToDataType>(value, precision_to, scale_from, scale_to, result);
+        else if constexpr (std::is_same_v<FromFieldType, Decimal256>)
+            return convertDecimalsImpl<DataTypeDecimal<Decimal256>, ToDataType>(value, precision_to, scale_from, scale_to, result);
+        else if constexpr (IsDataTypeNumber<FromDataType>)
+            return convertNumberToDecimalImpl<DataTypeNumber<FromFieldType>, ToDataType>(value, precision_to, scale_to, result);
+        else
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Convert from {} type to decimal type is not implemented.", typeid(value).name());
+    }
+
+    template <typename FromDataType, typename ToDataType>
+    requires(IsDataTypeNumber<FromDataType> && IsDataTypeDecimal<ToDataType>)
+    static bool convertNumberToDecimalImpl(
+        const typename FromDataType::FieldType & value,
+        UInt32 precision,
+        UInt32 scale,
+        typename ToDataType::FieldType & result)
+    {
+        using FromFieldType = typename FromDataType::FieldType;
+        double int_part = 0;
+        if constexpr (std::is_same_v<FromFieldType, Float32> || std::is_same_v<FromFieldType, BFloat16>)
+            int_part = static_cast<int>(value);
+        else if constexpr (std::is_same_v<FromFieldType, Float64>)
+            int_part = static_cast<long>(value);
+        else
+            int_part = value;
+
+        int int_part_digits = int_part == 0 ? 1 :
+                              int_part > 0 ? static_cast<int>(std::log10(int_part)) + 1
+                                           : static_cast<int>(std::log10(std::fabs(int_part))) + 1;
```

Comment:
I guess `std::log10` and `std::fabs` is too heavy for this function.  Maybe it is better:  
``` cpp
auto casted_int_part = static_cast<ToDataType::FieldType>(casted_int_part); 
bool overflow = casted_int_part > min_value && casted_int_value <  max_value;   
```

`min_value`/`max_value` is the minimum/maximum value which could be represented in `precision - scale` digits. They could be calculated outside for loop.